# What 'Hacker News' Posts Gets the Greatest Response?

This project explores a dataset of posts from the popular Hacker News website. Hacker News is a website were user-submitted stories are voted on by the sites users. Posts that make it to the top of the site can receive hundreds of thousands of clicks. The original dataset can be found [here](https://www.kaggle.com/hacker-news/hacker-news-posts).
The dataset used in this exercise is a modified version where all posts that received no comments have been removed. This leaves about 20,000 rows, down from the original 300,000 rows.

The project specifically looks at the difference between `Show HN` and `Ask HN` types of posts. The `Show HN` submits a project, product, or something else of interest for the Hacker News community to peruse. In `Ask HN` submissions a question is posed to the community. The project will attempt to determine the following:
- Do `Show HN` or `Ask HN` receive more comments on average?
- Does the average number of comments on a post depend on when it was posted?

## Modification & Exploration of the dataset

In [13]:
import datetime as dt
from pprint import pprint as pp
from csv import reader

opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)

In [14]:
print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [15]:
#Separating out the headers
headers = list(hn)
headers = headers[0]
hn = hn[1:]
print(headers, "\n")
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'] 

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


## 'Ask HN' & 'Show HN' Posts

We'll separates out the various type of 'Hacker News' post for comparison.

In [16]:
#Separating out Ask HN, Show HN & other posts
ask_posts = []
show_posts = []
other_posts = []

title = [row[1] for row in hn]
title = [row.lower() for row in title]

for row in hn:
    #convert headlines to lowercase
    title = row[1].lower()
    
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

In [17]:
#Checking the lists
print("Number of 'Ask HN' posts: ", len(ask_posts))
print("Number of 'Show HN' posts: ", len(show_posts))
print("Number of other posts: ", len(other_posts), "\n") 

print(*ask_posts[:5], sep = "\n")
print("\n")
print(*show_posts[:5], sep = "\n")

Number of 'Ask HN' posts:  1744
Number of 'Show HN' posts:  1162
Number of other posts:  17194 

['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55']
['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43']
['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14']
['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20']
['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']


['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03']
['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46']
['11590768', 'Show H

In [18]:
#average number of comments on Ask HN posts
total_ask_comments = 0
total_ask_comments = sum([int(row[4]) for row in ask_posts])
avg_ask_comments = total_ask_comments / len(ask_posts)

print(total_ask_comments)
print(avg_ask_comments)

24483
14.038417431192661


In [19]:
#find the average number of comments on Show HN posts
total_show_comments = 0
total_show_comments = sum([int(row[4]) for row in show_posts])
avg_show_comments = total_show_comments / len(show_posts)

print(total_show_comments)
print(avg_show_comments)

11988
10.31669535283993


### Comparison of number of comments

Exploration of the dataset shows that `Ask HN` posts receive about a third more comments than `Show HN` posts:

| Post Type | Total Comments | Avg. Comments per Article |
| ----------- | :-----------: | :----: |
| `Ask HN`  | 24483 | ~ 14 |
| `Show HN` | 11988 | ~ 10.5 |

Since `Ask HN` posts receive on average more comments the rest of the analysis will focus on that post type.

## 'Ask HN' posts & comments breakdown by hour

In [20]:
print(headers)

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


In [21]:
result_list = []

#list of 'Ask HN' post creation time and # of comments on that post
for post in ask_posts:
    result_list.append([post[6], int(post[4])])

print(result_list[:5])

[['8/16/2016 9:55', 6], ['11/22/2015 13:43', 29], ['5/2/2016 10:14', 1], ['8/2/2016 14:20', 3], ['10/15/2015 16:38', 17]]


In [22]:
counts_by_hour = {} #amount of Ask HN posts for each hour of the day
comments_by_hour = {} # amount of comments on all Ask HN posts for the hour

for result in result_list:
    
    date = result[0]
    date = dt.datetime.strptime(date, "%m/%d/%Y %H:%M")
    hour = dt.datetime.strftime(date, "%H")
    comment = result[1]
    
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = int(comment)
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += int(comment)
    
print(sorted(counts_by_hour.items()), "\n")
print(sorted(comments_by_hour.items()), "\n")

[('00', 55), ('01', 60), ('02', 58), ('03', 54), ('04', 47), ('05', 46), ('06', 44), ('07', 34), ('08', 48), ('09', 45), ('10', 59), ('11', 58), ('12', 73), ('13', 85), ('14', 107), ('15', 116), ('16', 108), ('17', 100), ('18', 109), ('19', 110), ('20', 80), ('21', 109), ('22', 71), ('23', 68)] 

[('00', 447), ('01', 683), ('02', 1381), ('03', 421), ('04', 337), ('05', 464), ('06', 397), ('07', 267), ('08', 492), ('09', 251), ('10', 793), ('11', 641), ('12', 687), ('13', 1253), ('14', 1416), ('15', 4477), ('16', 1814), ('17', 1146), ('18', 1439), ('19', 1188), ('20', 1722), ('21', 1745), ('22', 479), ('23', 543)] 



In [23]:
#average # of comments for post created during each hour of the day
avg_by_hour = []

for element in comments_by_hour:
    
    avg_by_hour.append([element, round(comments_by_hour[element] / counts_by_hour[element], 2)])
    
pp(sorted(avg_by_hour))


[['00', 8.13],
 ['01', 11.38],
 ['02', 23.81],
 ['03', 7.8],
 ['04', 7.17],
 ['05', 10.09],
 ['06', 9.02],
 ['07', 7.85],
 ['08', 10.25],
 ['09', 5.58],
 ['10', 13.44],
 ['11', 11.05],
 ['12', 9.41],
 ['13', 14.74],
 ['14', 13.23],
 ['15', 38.59],
 ['16', 16.8],
 ['17', 11.46],
 ['18', 13.2],
 ['19', 10.8],
 ['20', 21.52],
 ['21', 16.01],
 ['22', 6.75],
 ['23', 7.99]]


In [24]:
#Swap columns in order to sort according to highest average number of comments
swap_avg_by_hour = []

for row in avg_by_hour:
    temp = dt.datetime.strptime(row[0], "%H")
    time_format = dt.datetime.strftime(temp, "%H:%M")
    swap_avg_by_hour.append([time_format, row[1]])
    
sorted_swap = sorted(swap_avg_by_hour, key=lambda x: x[1],  reverse=True)

print('Top 5 Hours for Ask Posts Comments')
for row in sorted_swap[:5]:
    print(str.format('{hour}: {comments} average comments per post', hour=row[0], comments=row[1]))
    

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.8 average comments per post
21:00: 16.01 average comments per post


## Conclusion

The analysis reveal that the `Ask HN` type of posts are the most popular on 'Hacker News'. Further exploration of the number of comments on post for a particular hour of the day shows `15:00` to be the most active. In other words: a post has a greater chance of being commented on if it is visible on the front page of 'Hacker News' at around 15:00 (the time zone used in the dataset is **Eastern Standard Time in the US (EST, UTC - 5)**).